Input Settings Menu

zicklag opened the pull request:
It's working! 

Opening as a draft because I need to do a self-review, but it's functional. 🎉

![setings-menu-3](https://user-images.githubusercontent.com/25393315/179362033-d3b459cf-ad4b-4693-9d0f-eb8b633d449b.gif)

zicklag commented on Cargo.toml:
```diff
@@ -22,7 +23,7 @@ getrandom = { version = "0.2", features = ["js"] }
 bevy-inspector-egui = { version = "0.11", optional = true }
 bevy-inspector-egui-rapier = { version = "0.4", optional = true, features = ["rapier2d"] }
-leafwing_input_manager = { git = "https://github.com/Leafwing-Studios/leafwing-input-manager.git", rev = "e8bd6e0" }
+leafwing_input_manager = { git = "https://github.com/zicklag/leafwing-input-manager.git", branch = "backport-leafwing-dev-to-bevy-0.7" }
```
Note: Because I had to make a fix to leafwing, and they've already migrated to the pending bevy 0.5, there's no way around using my temporary fork for this anymore. I think we'll just have to wait 'till we migrate to Bevy 0.8 to switch this to a more permanent release.

zicklag commented:
Just did a more thorough review of my own code: added lots of comments, removed some unneeded code paths, and fixed a minor usability issue with button hovering and focusing. I think this is ready now.

odecay commented:
Gonna try to review this tomorrow. 

One thing to consider for a future pr is a "bind all" mode which when entered prompts for input to bind for the first Action on list, then immediately progresses to the next Action after pressing a key/button until all actions are bound to buttons.

64kramsystem commented:
Going to give a review as well tomorrow or the day after, as I'm about to travel :smile:

odecay commented on src/ui/main_menu.rs:
```diff
@@ -0,0 +1,768 @@
+    todo!("Implement sound settings UI");
+}
+
+/// Format an InpubKind as a user-facing string
```
nbd but InputKind

odecay reviewed (commented):
Looks good! I didn't have much meaningful feedback, noticed few small things that looked like spelling errors.

None of this should block this work or be included in this pr but some thoughts came up while looking it over.

I think maybe we will want to think more about what options players have for binding in the menu, I feel like maybe it should allow for selecting an input type and then only showing a map for that type instead of gamepad + keyboard 1/2 at the same time. 

This might also relate to how we expect bindings to work in multiplayer both in local multiplayer and networked multiplayer. Which player gets to set their bindings once the menu is triggered? Right now it sets for all players correct?

zicklag commented:
> noticed few small things that looked like spelling errors.

Good catch, I fixed those now. :+1: 

> I feel like maybe it should allow for selecting an input type and then only showing a map for that type instead of gamepad + keyboard 1/2 at the same time.

That's a great idea. We could definitely do that in a next iteration. It would be an easy update.

> This might also relate to how we expect bindings to work in multiplayer both in local multiplayer and networked multiplayer. Which player gets to set their bindings once the menu is triggered? Right now it sets for all players correct? 

Yep. Everybody player shares the same gamepad bindings, and there are the two keyboard bindings so that you can play two players with one keyboard.

If we added a way to rotate through which bindings you are actually setting, like you mentioned, then that would make it easier to set different gamepad bindings for player 1 and player 2.

I'd have done separate gamepad bindings for both gamepads if it weren't for the fact that I ran out of room. :) Selecting which binding you are setting first solves a lot of problems.

odecay commented:
> > noticed few small things that looked like spelling errors.
> 
> Good catch, I fixed those now. +1
> 
> > I feel like maybe it should allow for selecting an input type and then only showing a map for that type instead of gamepad + keyboard 1/2 at the same time.
> 
> That's a great idea. We could definitely do that in a next iteration. It would be an easy update.
> 
> > This might also relate to how we expect bindings to work in multiplayer both in local multiplayer and networked multiplayer. Which player gets to set their bindings once the menu is triggered? Right now it sets for all players correct?
> 
> Yep. Every player shares the same gamepad bindings, and there are the two keyboard bindings so that you can play two players with one keyboard.
> 
> If we added a way to rotate through which bindings you are actually setting, like you mentioned, then that would make it easier to set different gamepad bindings for player 1 and player 2.
> 
> I'd have done separate gamepad bindings for both gamepads if it weren't for the fact that I ran out of room. :) Selecting which binding you are setting first solves a lot of problems.

Yep all totally makes sense. Lets add an issue for next steps for input binding after we get this PR merged in. Feel free to merge whenever you feel its ready.

zicklag commented:
> Feel free to merge whenever you feel its ready.

Since nothing I'm doing next needs this I'll wait a day to let @64kramsystem look at it.

zicklag commented:
> Would you mind if I took that one?

Oh, not at all, go ahead. That one looked easy and I was trying to find what the most important things to get out before the next release were, but I'd rather you took it anyway.

> What are you planning on working on next btw?

Right now I'm working on a quick experiment to see if an idea I had regarding scripting is feasible or not.

I figured that if we have a solution for scripting, it might have an impact on how we program the players or items, so it'd be good to have some idea of what that might look like earlier rather than later.

It should just take a day or two to see if my idea has any merit, and it's kind of fun experimentation work so I figured I'd take a quick look as sort of a "break".

Then I'll come back to the #83 probably next.

odecay commented:
> > Would you mind if I took that one?
> 
> Oh, not at all, go ahead. That one looked easy and I was trying to find what the most important things to get out before the next release were, but I'd rather you took it anyway.
> 
> > What are you planning on working on next btw?
> 
> Right now I'm working on a quick experiment to see if an idea I had regarding scripting is feasible or not.
> 
> I figured that if we have a solution for scripting, it might have an impact on how we program the players or items, so it'd be good to have some idea of what that might look like earlier rather than later.
> 
> It should just take a day or two to see if my idea has any merit, and it's kind of fun experimentation work so I figured I'd take a quick look as sort of a "break".
> 
> Then I'll come back to the #83 if nobody else wants it, or I'll just work on the update binding menu.

okay perfect! scripting sounds awesome, followed along with the threads/issues on mun scripting/bevy_mod_scripting, were you thinking something like #107?

zicklag commented:
Replied here to make it more visible: https://github.com/fishfight/punchy/issues/107#issuecomment-1189447103

64kramsystem commented:
> > Feel free to merge whenever you feel its ready.
> 
> Since nothing I'm doing next needs this I'll wait a day to let @64kramsystem look at it.

Looks good to me (although I'm not very familiar with the areas :grimacing:) :smile:

64kramsystem reviewed (approved):
:+1:
